tests: updata the rbtree testcase

Zhaoningx opened the pull request:
Add a new testcase to:
        Verify some operations of rbree are running in logarithmic time.
        Verify an user defined structure contains rbtree node works.
        Verify "for each"style APIs work.


Notice: I changed the directory tests/unit to tests/benchmarks/.  Do you think this change is OK?




Signed-off-by: Ningx Zhao <ningx.zhao@intel.com>

andyross commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,205 @@
+
+static struct rbnode nodes[512];
+static int random_num[512] = {0};
+void random_generate(int size)
```
I'm completely confused about what's happening here.  Are you trying to shuffle the integers between 0 and 511 into the random_num array such that they appear in a random order, but only once each?  If so, I don't think this code works correctly.  It's also needlessly O(N^2) in the size of the array.  Maybe try this instead (untested, but the basic idea)?:

```c


void shuffle_random(void)
{
    int n = ARRAY_SIZE(random_num);

    /* Populate values */
    for (int i = 0; i < n; i++) {
        random_num[i] = i;
    }

    /* Randomize each by swapping with one of the remaining ones */
    for (int i = 0; i < n; i++) {
        int j = next_rand_mod(n - i);
        int tmp = random_num[i];

        random_num[i] = random_num[j];
        random_num[j] = tmp;
    }
}
```

Zhaoningx replied:
The code is very good, but there are a few same numbers,as you know, it can't insert a tree using two same data.  If I need to use disorderly ordinal data, I think I can use this code instead and fix it.

andyross commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,205 @@
+	struct rbnode *rx_node = tree->root;
+	struct rbnode *temp = NULL;
+
+	temp = (size == 0) ? rb_get_min(tree) : rb_get_max(tree);
```
"side", not size

Zhaoningx replied:
OK, Thanks.

andyross commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,205 @@
+
+	rb_insert(tree, &ins_nodes);
+	/*operations(insert and remove) are same times in the same node*/
+	while (1) {
```
General nitpick: a bug in the test logic or the tree will show up here as an infinite loop.  That **is** a test failure, but it's a very expensive one.  Though in practice this code and the test aren't actually buggy.

Zhaoningx replied:
I met this condition before. When I run the code, it will do as an infinite loop, until the stack broken. I think the reason is that inserting two same nodes into the tree, it will not work normally. So do you think how can I get the height of node in the tree instead this.:)

Zhaoningx replied:
I changed another method to implement that finding the height of the tree.

andyross commented on tests/unit/list/dlist.c:
```diff
@@ -167,7 +167,7 @@ static inline bool verify_tail_head(sys_dlist_t *list,
 
 /**
- * @brief Verify doubly linked list funtionalities
+ * @brief Verify doubly linked list functionalities
```
Please add new misspellings in a separate patch.

Zhaoningx replied:
OK, I will do this in a separate patch.

andyross reviewed (dismissed):
Some nitpicks, and a randomizer that I think needs to be fixed.  Broadly I like the design here very much, checking the tree height instead of clock times.

Zhaoningx commented on lib/os/rb.c:
```diff
@@ -93,7 +91,6 @@ static int find_and_stack(struct rbtree *tree, struct rbnode *node,
 			break;
 		}
 	}
-
```
OK, I have reverted it.

Zhaoningx commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,205 @@
+	}
+}
+
+static unsigned int next_rand_mod(unsigned int mod)
```
Yep, this code just copied from the existing rbtree test. Maybe it needn't do this, as you see, these random numbers are used as a index of the rbnode array. I think it's more representative to insert a tree by disorderly ordinal data than orderly ordinal data. And, the testcase must be compatible with all platforms, the API sys_rand32_get() is not effective for all. So maybe I changed the method that inserting a tree, just using some orderly ordinal data. Do you think the method is OK?

Zhaoningx commented:
Hi @andyross , thank you very much for your comments, I have changed my code, I think maybe I needn't disorder numbers, and I used the method of recurse to search height of the node of a tree. Then the height of the comparison node is less than or equal to the worst-case height of the equilibrium tree.Do you think this is effective?Adding some comments to point some issues if there are.Thanks again.

Zhaoningx commented:
Hi @andyross @andrewboie @nashif , would you like to review this PR again, thanks.

LeiW000 commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,205 @@
+	struct container_node *c_foreach_node;
+	struct rbnode *foreach_node;
+
+	memset(&test_tree_l, 0, sizeof(test_tree_l));
```
I agree with Andy. You don't need to zero test_tree_l.

Zhaoningx replied:
Hi @andyross @LeiW000 ,Thanks for your comments, after I double check this code, the 'test_tree_l' and 'node' is local variable and this sub-function just is called once time, so I removed "static",and zero-initialized manually. :)

LeiW000 replied:
Sorry, I was wrong.

Zhaoningx commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,205 @@
+
+	/*"for each" style iteration to verify node*/
+	RB_FOR_EACH(&test_tree_l, foreach_node) {
+		zassert_true(((struct container_node *)foreach_node)->value
```
Done, Thanks

Zhaoningx commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,205 @@
+	memset(&test_tree_l, 0, sizeof(test_tree_l));
+	test_tree_l.lessthan_fn = node_lessthan;
+	memset(tree_node, 0, sizeof(tree_node));
+	for (unsigned int i = 0; i < 10; i++) {
```
Done, thanks.

LeiW000 reviewed (approved):
It looks good to me.

tyrantlizard77 reviewed (commented):
Nice idea on using tree height to demonstrate O(log n)!

This generally looks good, please address recent feedback from others.

Zhaoningx commented:
> Nice idea on using tree height to demonstrate O(log n)!
> 
> This generally looks good, please address recent feedback from others.

Hi @andrewboie ,I think I have fixed all feedback about @andyross and @LeiW000 .

LeiW000 commented:
@andyross , @andrewboie , could you take look at the PR again? Please approve it if you have no other comment.

Zhaoningx commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,205 @@
+{
+	int count = 0;
+	static struct rbtree test_tree_l;
+	struct container_node tree_node[10];
```
Okey, I have fixed this code.

Commit:
tests: updata the rbtree testcase

add a new testcase to:
Verify some operations of rbree are running in
logarithmic time.
Verify an user defined structure contains rbtree node works.
verify "for each"style APIs work.

Signed-off-by: Ningx Zhao <ningx.zhao@intel.com>

Zhaoningx commented:
Hey @andyross  and @andrewboie , I have fixed all feedbacks, would you please review this PR, waiting for more comments.

ceolin commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,159 @@
+}
+
+uint32_t node_height;
+int search_height_recurse(struct rbnode *node, struct rbnode *finial_node)
```
I would change this algorithm, you don't need to rely on a global variable.
Just add this parameter to the function to keep the height track, something like:
```
int search_height_recurse(struct rbnode *node, struct rbnode *final_node, uint32_t current_height)
{
	if (node == NULL) {
		return -1;
	}

	if (node == final_node) {
		return current_height;
 	}

	current_height++;
	struct rbnode *ch = z_rb_child(node, tree.lessthan_fn(finial_node, node));
	return search_height_recurse(ch, final_node, current_height);
}
```

Note, I haven't check the algorithm above accuracy, just the idea. Then you don't need to keep that global, you just call this function with initial value 0.

Zhaoningx replied:
The code looks great, and I do a few modifications to make it can work better. I have used the code,thanks.

ceolin commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,159 @@
+{
+	int side = 0;
+
+	if (node == finial_node) {
```
s/finial_node/final_node/

ceolin commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,159 @@
+	/* zephyr can't do floating-point arithmetic,
+	 * so manual:dlog_N = 2log(TREE_SIZE) = 18
+	 */
+	int dlog_N = 18;
```
Make this value a constant close to TREE_SIZE definition, so if one needs to change that value will easily see that this value has to be changed as well.

Zhaoningx replied:
Yup, you are right,I also think it's not flexible, but it's hard to float point arithmetic in zephyr kernel.

Zhaoningx replied:
OK, I defined it below the macro TREE_SIZE.  BTW, Thanks for your all suggestions about this PR, if there are not anything improperly, would you approve the PR?

ceolin commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,159 @@
+	int dlog_N = 18;
+
+	node_height = 0;
+	search_height_recurse(root, test);
```
Please look my suggestion about this function. If you want to keep the current implementation there is not need to that function return an int if the result is ignored.

ceolin commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,159 @@
+ * logarithmic time
+ *
+ * @details
+ * Defining a tree, searching height of maxmum node, minmum node
```
typos

s/maxmum/maximum/
s/minmum/minimum/

ceolin commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,159 @@
+ *
+ * @details
+ * Defining a tree, searching height of maxmum node, minmum node
+ * and the node inserted or removed.And record those heights.
```
The period is making this sentence difficult to read

ceolin commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,159 @@
+void test_rbtree_perf(void)
+{
+	memset(&tree, 0, sizeof(tree));
+	memset(nodes, 0, sizeof(nodes));
```
nitpick, why are you initializing them ? Aren't they static global variables not used before ?

andrewboie replied:
agreed these are in BSS should already be zeroed

Zhaoningx replied:
Good to know, I have removed 'memset'.

ceolin commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,159 @@
+	static struct rbtree test_tree_l;
+	static struct container_node tree_node[10];
+	static struct container_node *c_foreach_node;
+	static struct rbnode *foreach_node;
```
Do these variables need to be static ?

Zhaoningx replied:
Double check it and I think it's local variables and do not need to return, so I don’t think need to add 'static'. And I have removed it, thanks.

ceolin commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,155 @@
+/* zephyr can't do floating-point arithmetic,
+ * so manual: dlog_N = 2log(TREE_SIZE) = 18
+ */
+const uint32_t dlog_N = 18;
```
`static` to limit this variable scope

Zhaoningx replied:
I totally agree with you, thanks.

ceolin commented on tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c:
```diff
@@ -0,0 +1,155 @@
+{
+	tree->lessthan_fn = node_lessthan;
+
+	for (int i = 0; i < size; i++) {
```
previous function you use an unsigned int in the for loop, here you are using int, please make it consistent.

ceolin reviewed (commented):
Some cosmetic issues, other than that looks great

Zhaoningx commented:
Hi @andyross @andrewboie @nashif , would you take a look of this PR, the PR has got two approvals, if there are not anything wrong, could you merge it, thanks.

tyrantlizard77 reviewed (approved):
LGTM.
@andyross wrote the rbtree code so will need his blessing.

Zhaoningx commented:
@andyross would you like to just spend a few time to review this PR?

Zhaoningx commented:
Hi @andyross , Could you take a look at this PR? Thanks!

nashif commented on tests/benchmarks/data_structure_perf/rbtree_perf/testcase.yaml:
```diff
@@ -0,0 +1,3 @@
+tests:
+  benchmark.data_structures:
+    tags: rbtree req-test
```
not sure what req-test is, please do not add arbitrary tags that are difficult to understand

Zhaoningx replied:
Hi @nashif , I have re-pushed the PR to modify the tags and make them more clear. PR link: https://github.com/zephyrproject-rtos/zephyr/pull/28110
